File: packages/python/src/vietflex/webgis/providers/google_maps.py
"""Google Maps URL coordinate parsing without network access."""

from __future__ import annotations

import re
from dataclasses import dataclass
from urllib.parse import parse_qs, urlparse


@dataclass(frozen=True, slots=True)
class LatLng:
    lat: float
    lng: float


_DIRECT = re.compile(
    r"^\s*(-?\d+(?:\.\d+)?)\s*[, ]\s*(-?\d+(?:\.\d+)?)\s*$"
)
_AT = re.compile(r"@(-?\d+(?:\.\d+)?),(-?\d+(?:\.\d+)?)(?:,|z|/)")
_DATA = re.compile(r"!3d(-?\d+(?:\.\d+)?)!4d(-?\d+(?:\.\d+)?)")


def parse_google_maps_url(value: str) -> LatLng | None:
    text = value.strip()
    if not text:
        return None

    for pattern in (_DIRECT, _AT, _DATA):
        match = pattern.search(text)
        if match:
            result = _normalize(float(match.group(1)), float(match.group(2)))
            if result:
                return result

    try:
        query = parse_qs(urlparse(text).query)
    except ValueError:
        return None

    for key in ("query", "destination", "q", "ll"):
        for candidate_text in query.get(key, []):
            match = _DIRECT.search(candidate_text)
            if match:
                result = _normalize(float(match.group(1)), float(match.group(2)))
                if result:
                    return result

    return None


def _normalize(lat: float, lng: float) -> LatLng | None:
    if abs(lat) > 90 or abs(lng) > 180:
        return None
    return LatLng(lat=lat, lng=lng)

File: packages/webgis/src/providers/googleMaps.ts
export interface LatLng {
  lat: number;
  lng: number;
}

export function parseGoogleMapsUrl(input: string): LatLng | null {
  const value = input.trim();
  if (!value) return null;

  const direct = parseDirectCoordinate(value);
  if (direct) return direct;

  const patterns = [
    /@(-?\d+(?:\.\d+)?),(-?\d+(?:\.\d+)?)(?:,|z|\/)/i,
    /!3d(-?\d+(?:\.\d+)?)!4d(-?\d+(?:\.\d+)?)/i
  ];

  for (const pattern of patterns) {
    const match = value.match(pattern);
    if (match) {
      const result = normalize(Number(match[1]), Number(match[2]));
      if (result) return result;
    }
  }

  try {
    const url = new URL(value);
    for (const key of ["query", "destination", "q", "ll"]) {
      const candidate = url.searchParams.get(key);
      if (!candidate) continue;
      const parsed = parseDirectCoordinate(candidate);
      if (parsed) return parsed;
    }
  } catch {
    return null;
  }

  return null;
}

export async function resolveGoogleMapsUrl(
  input: string,
  resolver: (url: string) => Promise<string>
): Promise<LatLng | null> {
  const parsed = parseGoogleMapsUrl(input);
  if (parsed) return parsed;
  return parseGoogleMapsUrl(await resolver(input));
}

function parseDirectCoordinate(value: string): LatLng | null {
  const match = value.match(
    /^\s*(-?\d+(?:\.\d+)?)\s*[, ]\s*(-?\d+(?:\.\d+)?)\s*$/
  );
  return match ? normalize(Number(match[1]), Number(match[2])) : null;
}

function normalize(lat: number, lng: number): LatLng | null {
  if (!Number.isFinite(lat) || !Number.isFinite(lng)) return null;
  if (Math.abs(lat) > 90 || Math.abs(lng) > 180) return null;
  return { lat, lng };
}
